chemprop/uncertainty/uncertainty_predictor.py

```diff
@@ -0,0 +1,842 @@
+        spectra_phase_mask: List[List[bool]] = None,
+    ):
+        self.test_data = test_data
+        self.models = models
+        self.scalers = scalers
+        self.dataset_type = dataset_type
+        self.loss_function = loss_function
+        self.uncal_preds = None
+        self.uncal_vars = None
+        self.uncal_confidence = None
+        self.individual_vars = None
+        self.num_models = num_models
+        self.uncertainty_dropout_p = uncertainty_dropout_p
+        self.dropout_sampling_size = dropout_sampling_size
+        self.individual_ensemble_predictions = individual_ensemble_predictions
+        self.spectra_phase_mask = spectra_phase_mask
+        self.train_class_sizes = None
+
+        self.raise_argument_errors()
+        self.test_data_loader = test_data_loader
+        self.calculate_predictions()
+
+    @property
+    @abstractmethod
+    def label(self):
+        """
+        The string in saved results indicating the uncertainty method used.
+        """
+
+    def raise_argument_errors(self):
+        """
+        Raise errors for incompatible dataset types or uncertainty methods, etc.
+        """
+
+    @abstractmethod
+    def calculate_predictions(self):
+        """
+        Calculate the uncalibrated predictions and store them as attributes
+        """
+
+    def get_uncal_preds(self):
+        """
+        Return the predicted values for the test data.
+        """
+        return self.uncal_preds
+
+    def get_uncal_vars(self):
+        """
+        Return the uncalibrated variances for the test data
+        """
+        return self.uncal_vars
+
+    def get_uncal_confidence(self):
+        """
+        Return the uncalibrated confidences for the test data
+        """
+        return self.uncal_confidence
+
+    def get_individual_vars(self):
+        """
+        Return the variances predicted by each individual model in an ensemble.
+        """
+        return self.individual_vars
+
+    def get_individual_preds(self):
+        """
+        Return the value predicted by each individual model in an ensemble.
+        """
+        return self.individual_preds
+
+    @abstractmethod
+    def get_uncal_output(self):
+        """
+        Return the uncalibrated uncertainty outputs for the test data
+        """
+
+
+class NoUncertaintyPredictor(UncertaintyPredictor):
+    """
+    Class that is used for predictions when no uncertainty method is selected.
+    Model value predictions are made as normal but uncertainty output only returns "nan".
+    """
+    @property
+    def label(self):
+        return "no_uncertainty_method"
+
+    def calculate_predictions(self):
+        for i, (model, scaler_list) in enumerate(
+            tqdm(zip(self.models, self.scalers), total=self.num_models)
+        ):
+            (
+                scaler,
+                features_scaler,
+                atom_descriptor_scaler,
+                bond_feature_scaler,
+            ) = scaler_list
```

Comment:
not sure on this change actually- but I'm generally of the opinion that tuples should be unpacked in the `for`-statement rather than the loop body. Good leaving this as is though

Reply:
The parentheses nesting is already a little bit high before then. I'd rather leave them as is.